gobblin-data-management/src/main/java/gobblin/data/management/convertion/hive/HiveSource.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (C) 2014-2016 LinkedIn Corp. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed
+ * under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ * CONDITIONS OF ANY KIND, either express or implied.
+ */
+package gobblin.data.management.convertion.hive;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import lombok.extern.slf4j.Slf4j;
+
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.hive.ql.metadata.Partition;
+import org.apache.hadoop.hive.ql.metadata.Table;
+import org.joda.time.DateTime;
+
+import com.google.common.base.Optional;
+import com.google.common.collect.Lists;
+import com.google.gson.Gson;
+
+import gobblin.configuration.SourceState;
+import gobblin.configuration.WorkUnitState;
+import gobblin.data.management.copy.hive.HiveDataset;
+import gobblin.data.management.copy.hive.HiveDatasetFinder;
+import gobblin.data.management.copy.hive.HiveUtils;
+import gobblin.dataset.IterableDatasetFinder;
+import gobblin.hive.HivePartition;
+import gobblin.hive.HiveTable;
+import gobblin.hive.metastore.HiveMetaStoreUtils;
+import gobblin.source.Source;
+import gobblin.source.extractor.Extractor;
+import gobblin.source.extractor.WatermarkInterval;
+import gobblin.source.workunit.WorkUnit;
+import gobblin.util.HadoopUtils;
+import gobblin.util.io.GsonInterfaceAdapter;
+import gobblin.util.reflection.GobblinConstructorUtils;
+
+
+/**
+ * <p>
+ * A {@link Source} that creates generic workunits for a hive table or a hive partition.
+ * </p>
+ * <ul>
+ *  <li>It uses the {@link HiveDatasetFinder} to find all hive tables and partitions
+ *  <li>The update time of a hive {@link Table} or a hive {@link Partition} if found using {@link HiveUnitUpdateProvider}
+ *  <li>The update time from the previous run is used as previous hive watermark.{@link HiveSourceWatermarker} is
+ *  used to get previous hive watermarks
+ * </ul>
+ *
+ *{@link WorkUnit}s are created if the previous high watermark of a {@link Partition}
+ * or a {@link Table} are lower than the latest update time.
+ *
+ * <p>
+ * The {@link WorkUnit}s contain a serialized json of the {@link HiveTable} or {@link HivePartition} at {@value #HIVE_UNIT_SERIALIZED_KEY}
+ * This is later deserialized by the extractor.
+ * </p>
+ */
+@Slf4j
+@SuppressWarnings("rawtypes")
```

Comment:
You can use `Source<Schema, QueryBasedHiveConversionEntity>` instead of raw type.